### ISSUE TITLE
Consider owning @types/cheerio type definitions

### ISSUE THREAD

**Issue by matthewmueller** [cheeriojs/cheerio#1471]
### Problem

In the early days, jQuery familiarity was one of the main attractions to Cheerio. Nowadays it may be more of a hindrence.

For example, it's confusing that Cheerio's `$(selector).each(fn)` has a different argument order than ES5's `forEach`.

```js
const fruits = [];

$('li').each(function(i, elem) {
  fruits[i] = $(this).text();
});

fruits.join(', ');
//=> Apple, Orange, Pear
```

### Potential Solution

One way to reduce the friction is to have autocomplete. We currently support some autocomplete via JSDoc but the type definitions would take it a step further.

Some wonderful folks have been maintaining [@types/cheerio](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3e0007a3a88eca10f61fcb55d78dd66e323f24c6/types/cheerio/index.d.ts). Installing this extra package is quite obvious for Typescript users but less obvious for Javascript developers. 

We could make this simpler by hosting those types within the Cheerio repository. This would be one less thing to install and the type definitions would get picked up for Javascript users.

### Downsides

- The maintenance burden is now on the Cheerio maintainers. 
- Any other downsides?

**Comment by bbqsrc** [comment]
Writing a scraper, as I am wont to do at least once per year, and straight up came here to ask why it wasn't hosted in the repo yet.

Timing was good, it seems. :smile:

**Comment by fb55** [comment]
@AzSiAz Afaict you seem to be the person most on top of the current type definitions (thank you or that!). Any concerns about us pulling the type definitions into the cheerio repo? Anyone else you think we should loop in?

**Comment by AzSiAz** [comment]
@fb55 Thanks for asking, glad you're a considering owning the type definition :smile:

Maybe the other typing maintainer so they know if they need to make modification but I think most of them don't participate anymore so do as you see fit on that front

The only concern coming to mind is to be careful about the node definition included in the current types, you may have to remove/"polyfill" it instead since many people seem to have a problem with it, and it's not really a concern but there is also a PR open [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46006): you might want to check with the author to get change concerning cheerio integrated 

Also, don't forget to delete types on definitelyTyped and update types depending on cheerio, might be a pain thought^^

**Comment by fb55** [comment]
Thanks a lot @AzSiAz, this is very helpful.

@paulmelnikow Should we use your PR as a basis for pulling in types? Any potential downsides that we should expect?

Also cc @blittle @wmaurer @umarniz @LiJinyao @chennakrishna8 @nwtgck @privatenumber — how do you feel about cheerio pulling in the type definition? Any concerns?

**Comment by paulmelnikow** [comment]
Yes, I think it is a great idea to pull in the types :+1:.

Cheerio does have a lot of dependencies and I don’t know how it will work when other DefinitelyTyped packages need to depend on types in cheerio. (However I bet DT has a solution for that!)

I would suggest, since my PR is making a pretty big change, maybe it’s better to get it merged and shipped from DT (assuming that can be done timely). Then you can pull in something that’s neatly identical to the latest shipped version.

**Comment by matthewmueller** [comment]
@paulmelnikow thanks for working on this!

> Cheerio does have a lot of dependencies and I don’t know how it will work when other DefinitelyTyped packages need to depend on types in cheerio. However I bet DT has a solution for that!

To clarify, do you mean that `@types/cheerio` has a lot of dependencies like `@types/chai-enzyme` that import `@types/cheerio`? Or that `@types/cheerio` depends on a lot of additional `@types/*` ?

Personally, I think we can copy these in and let DT sort it out on their end. They might need to do something like:

```
// inside @types/chai-enzyme
import type { Cheerio } from 'cheerio'
```

Another thing we'd need is to make sure we have are tests for our type definitions. 

@paulmelnikow how does [this test in DT](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b488e266a33a4965ec373d974b313eecd68bcd39/types/cheerio/cheerio-tests.ts) work? Does it run assertions?

I've also seen https://github.com/SamVerschueren/tsd be used to test types.

**Comment by paulmelnikow** [comment]
> @paulmelnikow how does [this test](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b488e266a33a4965ec373d974b313eecd68bcd39/types/cheerio/cheerio-tests.ts) work in DT? Does it run assertions?

Yea, it's checking that the provided code compiles successfully, or in the case of `$ExpectType` that the expression has the type it's supposed to have.

We use tsd to [check the type definitions in Shields](https://github.com/badges/shields/blob/master/badge-maker/index.test-d.ts). Would be a good way to verify the types when they're moved to the Cheerio repo.

**Comment by paulmelnikow** [comment]
I suppose we should hold off on deprecating the DT types until the 1.0.0 release is made.

**Comment by fb55** [comment]
Fixed in #1491
